Improve ZLayer how-to documentation (#4156)
zio/zio: docs/howto/use_modules_and_layers.md

 - `ZLayer.fromEffect` to lift a `ZIO` effect to a layer requiring the effect environment
 - `ZLayer.fromAcquireRelease` for a layer based on resource acquisition/release. The idea is the same as `ZManaged`
 - `ZLayer.fromService` to build a layer from a service
 - `ZLayer.fromServices` to build a layer from a number of required services
 - `ZLayer.identity` to express the requirement for a layer
 - `ZIO.toLayer` or `ZManaged.toLayer` to construct a layer from an effect

Where it makes sense, these methods have also variants to build a service effectfully (suffixed by `M`), resourcefully (suffixed by `Managed`), or to create a combination of services (suffixed by `Many`).

We can compose `layerA` and `layerB`  _horizontally_ to build a layer that has the requirements of both layers, to provide the capabilities of both layers, through `layerA ++ layerB`

We can also compose layers _vertically_, meaning the output of one layer is used as input for the subsequent layer to build the next layer, resulting in one layer with the requirement of the first, and the output of the second layer: `layerA >>> layerB`.

> When doing this, we must feed all the inputs to make the code compile, but we can still define part of the input as the input of the final layer using `ZLayer.identity`. 

Reviewer: ```suggestion
When doing this, the first layer must output all the services required by the second layer, but we can defer creating some of these services and require them as part of the input of the final layer using `ZLayer.identity`. 
```

Notice that in `layer` the dependency of `UserRepo` on `Connection` has been "hidden" and is no longer expressed in the type signature. From the perspective of a caller, `layer` just outputs a `UserRepo` and requires no inputs. The caller does not need to be concerned with the internal implementation details of how the `UserRepo` is constructed.

> To hide only some inputs we need to explicitly define left overs

Reviewer: ```suggestion
To provide only some inputs we need to explicitly define what inputs still need to be provided:
```

```scala mdoc:silent
trait Configuration

val userRepoWithConfig: ZLayer[Has[Configuration] with Has[Connection], Nothing, UserRepo] = 
  ZLayer.succeed(new Configuration{}) ++ postgresLayer
val partialLayer: ZLayer[Has[Configuration], Nothing, UserRepo] = 
  (ZLayer.identity[Has[Configuration]] ++ connection) >>> userRepoWithConfig
``` 

> In this example the dependency on `Connection` remains hidden, but `Configuration` is still required by `partialLayer`.

Reviewer: ```suggestion
In this example the requirement for a `Connection` has been satisfied, but `Configuration` is still required by `partialLayer`.
```
